client/api_object.go

```diff
@@ -262,6 +298,146 @@ func (c *client) PutObject(ctx context.Context, bucketName, objectName string, o
 	return nil
 }
 
+// UploadSegmentHook is for testing usage
+type uploadSegmentHook func(id int) error
+
+var UploadSegmentHooker uploadSegmentHook = DefaultUploadSegment
+
+func DefaultUploadSegment(id int) error {
+	return nil
+}
+
+func (c *client) putObjectResumable(ctx context.Context, bucketName, objectName string, objectSize int64,
+	reader io.Reader, opts types.PutObjectOptions,
+) (err error) {
+
+	if err := c.headSPObjectInfo(ctx, bucketName, objectName); err != nil {
+		log.Error().Msg(fmt.Sprintf("fail to head object %s , err %v ", objectName, err))
+		return err
+	}
+
+	offset, err := c.GetObjectResumableUploadOffset(ctx, bucketName, objectName)
+	if err != nil {
+		return err
+	}
+
+	// Total data read and written to server. should be equal to
+	// 'size' at the end of the call.
+	var totalUploadedSize int64
+
+	// Calculate the optimal parts info for a given size.
+	totalPartsCount, partSize, _, err := c.SplitPartInfo(objectSize, opts.PartSize)
+	if err != nil {
+		return err
+	}
+
+	// Part number always starts with '1'.
+	partNumber := 1
+	startPartNumber := int(offset/opts.PartSize + 1)
+
+	// Create a buffer.
+	buf := make([]byte, partSize)
+	complete := false
+
+	// skip sucess segment TODO(chris): seek ?
+	for partNumber < startPartNumber {
+		length, rErr := utils.ReadFull(reader, buf)
+		if rErr == io.EOF && partNumber > 1 {
+			break
+		}
+		// Increment part number.
+		log.Debug().Msg(fmt.Sprintf("skip partNumber:%d, length:%d", partNumber, length))
+		// Save successfully uploaded size.
+		totalUploadedSize += int64(length)
+		partNumber++
+	}
+
+	for partNumber <= totalPartsCount {
+		if partNumber == totalPartsCount {
+			complete = true
+		}
+		// hook for test
```

Comment:
Do not put test code/comment to the main process, its not a good practice. We can use hooks as one of the functions provided by the interface and then use them for testing.